Support global AWS endpoint overrides via configuration values

scprek opened the pull request:
Resurrecting https://github.com/micronaut-projects/micronaut-aws/pull/831. Not sure if there's an issue doing this at in the parent or client builder like in the S3 ticket https://github.com/micronaut-projects/micronaut-aws/pull/1420

fixes: https://github.com/micronaut-projects/micronaut-aws/issues/790

scprek commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/ClientConfigurationProperties.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.net.URI;
+
+public abstract class ClientConfigurationProperties extends AWSConfiguration {
```
Moved comments from S3ConfigurationProperties here

alvarosanchez reviewed (changes requested):
@scprek thank you very much to move this forward!

Please ensure that `./gradlew check` passes locally. Currently the build failed because there are some new files that need the license header. You can apply it automatically running `./gradlew spotlessApply`.

scprek commented:
@alvarosanchez I was trying to update the adocs too and somehow use aws.clients.*.endpoint-override. I did this by adding the  @EachProperty annotation. However this then caused the client specific configs to be aws.clients.*.s3.endpoint-override
vs staying aws.clients.s3.endpoint-override

scprek commented:
<img width="1390" alt="Screen Shot 2022-09-08 at 7 31 59 AM" src="https://user-images.githubusercontent.com/2783947/189122469-fac13c90-8da8-44e1-aaaa-46731f52da66.png">
<img width="1053" alt="Screen Shot 2022-09-08 at 7 32 08 AM" src="https://user-images.githubusercontent.com/2783947/189122499-f823aa2e-e439-412e-85d7-15c96530b23b.png">

scprek commented:
@alvarosanchez Everything I tried wrt .adoc auto documentation with configuration properties didn't work or would require lots of changes. So I manually added which is unfortunate

https://github.com/scprek/micronaut-aws/blob/eade027df144890056960a6fb51c45c08104fe62/src/main/docs/guide/sdkv2/serviceClients.adoc

scprek commented:
<img width="1567" alt="Screen Shot 2022-09-12 at 9 59 08 AM" src="https://user-images.githubusercontent.com/2783947/189687544-c74a4b2b-2cc7-49ce-bb56-9fcc5f5bcb02.png">

scprek commented:
<img width="1560" alt="Screen Shot 2022-09-12 at 10 00 45 AM" src="https://user-images.githubusercontent.com/2783947/189687931-c0776e44-2891-4bce-b1c2-8fca6133bac8.png">

scprek commented:
I also noticed that there's a config file for secrets manager in a different directory... so now we have configurations in different places https://github.com/micronaut-projects/micronaut-aws/blob/master/aws-secretsmanager/src/main/java/io/micronaut/aws/secretsmanager/SecretsManagerConfigurationProperties.java

scprek commented:
I seem to have failed `aws-sdk-v2:japiCmp` and not sure how to complete this?
<img width="1608" alt="Screen Shot 2022-09-12 at 3 58 41 PM" src="https://user-images.githubusercontent.com/2783947/189757415-b12bca0b-2955-4d8b-8e8c-bf045366ee96.png">

scprek commented:
@n0tl3ss @melix @alvarosanchez I made changes to use `@EachProperty` for common configurations

```java
@EachProperty("services")
public class AWSServiceConfiguration extends AWSConfiguration {
```

and one for service specific configs to extend
```java
@ConfigurationProperties(AWSServiceConfigurationProperties.SERVICE_PREFIX)
public class AWSServiceConfigurationProperties extends AWSConfiguration {
```

and would love some feedback. There are currently various spots in the docs for different AWS services(secrets manager, cloudwatchlogging, which I moved?) and them also being nested under sdk2?

<img width="2239" alt="Screen Shot 2022-09-13 at 8 56 57 PM" src="https://user-images.githubusercontent.com/2783947/190041323-3e205fc4-9705-46dc-824f-4b0234ad037f.png">
<img width="2230" alt="Screen Shot 2022-09-13 at 8 56 52 PM" src="https://user-images.githubusercontent.com/2783947/190041330-6c00e53c-20e4-45c2-bfbe-c66a6be3c834.png">

scprek commented on aws-sdk-v2/src/test/java/io/micronaut/aws/sdk/v2/RekognitionConfigurationProperties.java:
```diff
@@ -0,0 +1,24 @@
+
+//tag::class[]
+@ConfigurationProperties(RekognitionClient.SERVICE_NAME)
+public class RekognitionConfigurationProperties  {
```
Most ConfigurationProperties classes I made are empty since the `@EachProperty` change, I did include what was done in the S3ConfigurationProperties. Not sure if every single service needs one of these would have Builders

alvarosanchez replied:
So if they leverage underlying SDK builders, then I think it is justified to have them. Otherwise it's pointless to keep them, IMO

scprek replied:
Sounds good that’s what I was thinking too

scprek commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/cloudwatchlogs/CloudWatchLogsConfigurationProperties.java:
```diff
@@ -0,0 +1,31 @@
+ * @since 3.9.0
+ */
+@ConfigurationProperties(CloudWatchLogsClient.SERVICE_NAME)
+public class CloudWatchLogsConfigurationProperties extends AWSServiceConfigurationProperties {
```
These are currently unused now that I changed over to `@EachProperty`, not sure if they all need builders

alvarosanchez replied:
Yes, no need to have them as far as I can tell. Please remove.

scprek commented:
@alvarosanchez not sure how to address the build issues. I've rebased to master and have run `./gradlew check` along with individual tasks that said they failed locally and everything passes using java 11. Do I have to recompile against java 8 and test locally or is there some trick to run locally?


Sonar failure
```
* What went wrong:
[1462](https://github.com/micronaut-projects/micronaut-aws/actions/runs/3049664265/jobs/4918538490#step:7:1463)
Execution failed for task ':sonarqube'.
[1463](https://github.com/micronaut-projects/micronaut-aws/actions/runs/3049664265/jobs/4918538490#step:7:1464)
> Project not found. Please check the 'sonar.projectKey' and 'sonar.organization' properties, the 'SONAR_TOKEN' environment variable, or contact the project administrator
```

melix commented:
AFAIK Sonar won't run on community PRs.

alvarosanchez commented:
> AFAIK Sonar won't run on community PRs.

That's correct.

However, @scprek there are still other failures:

```
FAILURE: Executed 120 tests in 1m 33s (102 failed)
> Task :aws-sdk-v2:test FAILED
```

Note that I still have not reviewed in detail this PR.

scprek commented:
Understood but I'm unable to recreate those errors locally. I was asking if there is some micronaut build trick for testing against Java 8. As that's what's failing right?

alvarosanchez commented:
Simply use a JDK 8 locally

scprek commented:
> Simply use a JDK 8 locally

🤣 

Thanks, I wasn't sure if there was some target flag that needed to change somewhere.

scprek commented:
Fixed the java 8 issue https://bugs.openjdk.org/browse/JDK-8141508

scprek commented:
@alvarosanchez I updated to since 3.10. I'm not sure if having the since is required/expected or not.

scprek commented:
@alvarosanchez is there anyone else I should ask to review?

alvarosanchez commented:
No need to. Please be patient, it's on my list, and I will get to this soon.

scprek commented:
Just checking in @alvarosanchez

alvarosanchez commented:
@scprek I appreciate your patience with this, the notifications about this got buried within a bunch of others. I'm reviewing this now

alvarosanchez commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/AwsClientFactory.java:
```diff
@@ -60,10 +68,13 @@ protected AwsClientFactory(AwsCredentialsProviderChain credentialsProvider, AwsR
                 .httpClient(httpClient)
                 .region(regionProvider.getRegion())
                 .credentialsProvider(credentialsProvider);
+        // Do not replace with method reference - https://bugs.openjdk.org/browse/JDK-8141508
```
```suggestion
        // Do not replace with method reference - https://bugs.openjdk.org/browse/JDK-8141508
        //noinspection Convert2MethodRef        
```

So that, at least IDEA, stops highlighting the line.

Please add it to the other places too

alvarosanchez commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/AwsClientFactory.java:
```diff
@@ -39,16 +42,21 @@
     protected final AwsCredentialsProviderChain credentialsProvider;
     protected final AwsRegionProviderChain regionProvider;
+    @Nullable
+    protected final AWSServiceConfiguration configuration;
```
If it's nullable then it shouldn't be final no?

alvarosanchez replied:
Resolving as I see that `regionProvider` is in the same situation

alvarosanchez commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/cloudwatchlogs/CloudwatchLogsClientFactory.java:
```diff
@@ -45,9 +47,11 @@ public class CloudwatchLogsClientFactory extends AwsClientFactory<CloudWatchLogs
      */
-    protected CloudwatchLogsClientFactory(AwsCredentialsProviderChain credentialsProvider, AwsRegionProviderChain regionProvider) {
-        super(credentialsProvider, regionProvider);
+    protected CloudwatchLogsClientFactory(AwsCredentialsProviderChain credentialsProvider, AwsRegionProviderChain regionProvider, ApplicationContext context) {
+        super(credentialsProvider, regionProvider, context.findBean(
+            AWSServiceConfiguration.class, Qualifiers.byName(CloudWatchLogsClient.SERVICE_NAME)).orElse(null));
     }
```
I believe you can receive directly the `@Named(CloudWatchLogsClient.SERVICE_NAME) AWSServiceConfiguration` instead of the application context

alvarosanchez commented on aws-sdk-v2/src/test/groovy/io/micronaut/aws/sdk/v2/service/CloudWatchLogsClientSpec.groovy:
```diff
@@ -1,24 +1,40 @@
+        client.serviceName() == serviceName()
+    }
+
+    void "it can override the endpoint"() {
```
Perhaps you could move this test to `ServiceClientSpec` and add abstract methods for the subclasses to provide the actual `SdkClient` client and asyncClient instances.

alvarosanchez commented on aws-secretsmanager/src/test/groovy/io/micronaut/aws/secretsmanager/SecretsManagerConfigurationSpec.groovy:
```diff
@@ -5,7 +5,7 @@ class SecretsManagerConfigurationSpec  extends ApplicationContextSpecification {
     Map<String, Object> getConfiguration() {
         super.configuration + [
-                'aws.secretsmanager.enabled': false
+                'aws.services.secretsmanager.enabled': false
```
This is a breaking change. I understand the reason why you changed it (agree that consistency is good), but it being a breaking change would require a new major release of micronaut-aws.

I suggest keeping them as they are for now.

sdelamo commented:
@scprek thanks for the PR and the patience. I am going to fork your PR and try to implement this in a non breaking change.

scprek commented:
Thanks so much @alvarosanchez ans @sdelamo . With respect to reverting the services config breaking change, does that mean the docs will need to be updated to have endpoint override mentioned only where it applies vs what I did to make it global?

alvarosanchez commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/s3/S3ConfigurationProperties.java:
```diff
@@ -50,8 +50,10 @@ public Builder getBuilder() {
     /**
      * @return The endpoint with which the AWS SDK should communicate
      * @since 3.6.2
+     * Use configuration {@literal aws.services.s3.endpoint-override} instead.
```
```suggestion
     * @deprecated Use configuration {@literal aws.services.s3.endpoint-override} instead.
```

alvarosanchez commented on aws-sdk-v2/src/main/java/io/micronaut/aws/sdk/v2/service/s3/S3ConfigurationProperties.java:
```diff
@@ -60,7 +62,9 @@ public URI getEndpointOverride() {
      * Provide a URI to override the endpoint with which the AWS SDK should communicate. Optional. Defaults to `null`.
      * @param endpointOverride The endpoint with which the AWS SDK should communicate
      * @since 3.6.2
+     * Use configuration {@literal aws.services.s3.endpoint-override} instead.
```
```suggestion
     * @deprecated Use configuration {@literal aws.services.s3.endpoint-override} instead.
```

alvarosanchez reviewed (changes requested):
Minor issues in Javadoc, but other than that, good to go!

scprek commented:
I'll generate the adocs and make sure they look ok

scprek commented:
@sdelamo @alvarosanchez are we ok with the docs just saying this now? Vs before https://github.com/micronaut-projects/micronaut-aws/pull/1456/files#issuecomment-124613464... the breaking change

Do we now have both `aws.s3.endpoint-override` and `aws.services.s3.endpoint-override` for example? The docs are confusing now.

<img width="984" alt="Screen Shot 2022-12-15 at 11 03 34 AM" src="https://user-images.githubusercontent.com/2783947/207922781-582679e1-ffa6-48b1-939d-d1fd470c0ae5.png">

alvarosanchez commented:
Can we not get rid of the `services` prefix altogether?

scprek commented:
> Can we not get rid of the `services` prefix altogether?

Yeah So I added the s3 prefix so I could have the adocs auto generate it for each service. But since we dropped that, wondering if we should revert back to old paths and just have manual docs.

sdelamo commented:
> Do we now have both `aws.s3.endpoint-override` and `aws.services.s3.endpoint-override` for example? The docs are can be confusing now

Unfortunately, we have to keep `aws.s3.endpoint-override` as a valid option. Otherwise, we will break users' applications .

sdelamo commented:
> > Can we not get rid of the `services` prefix altogether?
> 
> Yeah So I added the service prefix so I could have the adocs auto generate it for each service. But since we dropped that, wondering if we should revert back to old paths and just have manual docs.

I think it is better to use `aws.services`. We have many places where we already have configuration with `aws.`. E.g. `aws.client`.
https://micronaut-projects.github.io/micronaut-aws/latest/guide/configurationreference.html

I think it is safer and easier to reason about if we stick with `@EachProperty("aws.services") instead of `@EachProperty("aws")`. If we used @EachProperty("aws")` It will create a bean `AwsServiceConfiguration` with name qualifier `client` and we don't want that.

scprek commented:
Yeah I like the `aws.services` for those reasons as well. I went ahead and just added a clarifying note to the s3 docs for now.

<img width="1025" alt="Screen Shot 2022-12-15 at 11 26 41 AM" src="https://user-images.githubusercontent.com/2783947/208102157-2a17abc7-311f-4454-8aba-62d195a302c3.png">

scprek commented:
@sdelamo @alvarosanchez good to go?

melix commented:
Is this a breaking change which should go in `4.0.0` instead?

scprek commented:
> Is this a breaking change which should go in `4.0.0` instead?

I believe @sdelamo did work to make it non-breaking and moved that to this PR correct? https://github.com/micronaut-projects/micronaut-aws/pull/1531

Currently for s3 we support both `aws.s3.endpoint-override` and `aws.services.*.endpoint-override` and I think the constructors are backwards compatible.

melix commented:
Ok great!
